src/tribol/geom/NodalNormal.cpp

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Tribol Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (MIT)
+
+#include "NodalNormal.hpp"
+
+#include "tribol/utils/Math.hpp"
+
+#ifdef TRIBOL_USE_ENZYME
+#include "tribol/common/Enzyme.hpp"
+#endif
+
+namespace tribol {
+
+void ElementAvgNormal::Compute( MeshData& mesh )
+{
+  if ( mesh.numberOfElements() == 0 ) {
+    return;
+  }
+
+  mesh.allocateNodalNormals();
+
+  auto mesh_view = mesh.getView();
+  // check to make sure face normals have been computed with
+  // a call to computeFaceData
+  SLIC_ERROR_IF( !mesh_view.hasElementNormals(), "MeshData::computeNodalNormals: required face normals not computed." );
+
+  // loop over elements
+  for ( int i = 0; i < mesh_view.numberOfElements(); ++i ) {
+    // loop over element nodes
+    for ( int j = 0; j < mesh_view.numberOfNodesPerElement(); ++j ) {
+      // SRW: note the connectivity array must be local to the mesh for indexing into
+      // the mesh nodal normal array. If it is not, then nodeId will access some other
+      // piece of memory and there may be a memory issue when numFaceNrmlsToNodes is deleted
+      // at the end of this routine.
+      int nodeId = mesh_view.getGlobalNodeId( i, j );
+      for ( int d = 0; d < mesh_view.spatialDimension(); ++d ) {
+        mesh_view.getNodalNormals()( d, nodeId ) += mesh_view.getElementNormals()( d, i );
+      }
+
+    }  // end loop over element nodes
+
+  }  // end loop over elements
+
+  // normalize the nodal normals
+  if ( mesh_view.spatialDimension() == 3 ) {
+    for ( int i = 0; i < mesh_view.numberOfNodes(); ++i ) {
+      RealT mag = magnitude( mesh_view.getNodalNormals()( 0, i ), mesh_view.getNodalNormals()( 1, i ),
+                             mesh_view.getNodalNormals()( 2, i ) );
+      mesh_view.getNodalNormals()( 0, i ) /= mag;
+      mesh_view.getNodalNormals()( 1, i ) /= mag;
+      mesh_view.getNodalNormals()( 2, i ) /= mag;
+    }
+  } else {
+    for ( int i = 0; i < mesh_view.numberOfNodes(); ++i ) {
+      RealT mag = magnitude( mesh_view.getNodalNormals()( 0, i ), mesh_view.getNodalNormals()( 1, i ) );
+      mesh_view.getNodalNormals()( 0, i ) /= mag;
+      mesh_view.getNodalNormals()( 1, i ) /= mag;
+    }
+  }
+}
+
+#ifdef TRIBOL_USE_ENZYME
+
+void VertexAvgNormal::Compute( MeshData& mesh )
```

Comment:
If this is within the enzyme ifdef should we follow the convention established elsewhere and call the class VertexAvgNormalEnzyme? Or append enzyme to the new name if we change it? Otherwise it is confusing who can use this class when and where? Is there any reason to compute the normal this way when NOT using enzyme? What if we use mortar with the hand coded approximate tangent stiffness, but want to change which nodal normal calculation we use?

Reply:
moved ifdef to a smaller scope so this can be used without Enzyme.